deepspeed/checkpoint/utils.py

```diff
@@ -29,3 +30,31 @@ def get_layer_ckpt_name_for_rank(base_folder, layer_id, tp_rank):
     ckpt_file = f'{layer_id}-model_{tp_rank:02d}{MODEL_FILE_SUFFIX}'
     ckpt_path = os.path.join(base_folder, ckpt_file)
     return ckpt_path
+
+
+# We pass cloned tensors to torch.save() to avoid checkpoint bloat that occurs when torch.save()
+# saves the underlying storage rather than the slice of the storage corresponding to individual tensors.
+# This is a problem in DeepSpeed because we often allocate tensors using slices of large flattened buffers.
+# Tensor cloning helps to avoid this problem because the storage of cloned tensors are closer to the true size.
+# It is expected that the garbage collector will reclaim the cloned tensor storage to avoid memory bloat.
+# See https://pytorch.org/docs/stable/notes/serialization.html#preserve-storage-sharing
+def clone_tensors_for_torch_save(item, device=torch.device('cpu')):
+    """
+    Returns a copy of input with all enclosed tensors replaced by clones on specified device.
+    Works on individual tensors, and tensors contained/nested in lists, tuples, and dicts.
+    Parameters:
+        item: tensor to clone or (possibly nested) container of tensors to clone.
+        device: target device
+    Returns:
+        copy of input with cloned tensors on target device
+    """
+    if torch.is_tensor(item):
+        return item.clone().to(device)
```

Comment:
probably need to call `.detach()` as well just in case.

Reply:
```suggestion
        return item.clone().detach().to(device)
```

Reply:
may be let's not do anything and let the user do any `detach` if they need it to?